streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredVersionedKeyValueStore.java

```diff
@@ -0,0 +1,227 @@
+     * {@link VersionedKeyValueStore} interface itself.
+     */
+    private class MeteredVersionedKeyValueStoreInternal
+        extends MeteredKeyValueStore<K, ValueAndTimestamp<V>>
+        implements TimestampedKeyValueStore<K, V> {
+
+        private final VersionedBytesStore inner;
+
+        MeteredVersionedKeyValueStoreInternal(final VersionedBytesStore inner,
+                                              final String metricScope,
+                                              final Time time,
+                                              final Serde<K> keySerde,
+                                              final Serde<ValueAndTimestamp<V>> valueSerde) {
+            super(inner, metricScope, time, keySerde, valueSerde);
+            this.inner = inner;
+        }
+
+        @Override
+        public void put(final K key, final ValueAndTimestamp<V> value) {
+            super.put(
+                key,
+                // versioned stores require a timestamp associated with all puts, including tombstones/deletes
+                value == null
+                    ? ValueAndTimestamp.makeAllowNullable(null, context.timestamp())
+                    : value
+            );
+        }
+
+        public ValueAndTimestamp<V> get(final K key, final long asOfTimestamp) {
+            Objects.requireNonNull(key, "key cannot be null");
+            try {
+                return maybeMeasureLatency(() -> outerValue(inner.get(keyBytes(key), asOfTimestamp)), time, getSensor);
+            } catch (final ProcessorStateException e) {
+                final String message = String.format(e.getMessage(), key);
+                throw new ProcessorStateException(message, e);
+            }
+        }
+
+        public ValueAndTimestamp<V> delete(final K key, final long timestamp) {
+            Objects.requireNonNull(key, "key cannot be null");
+            try {
+                return maybeMeasureLatency(() -> outerValue(inner.delete(keyBytes(key), timestamp)), time, deleteSensor);
+            } catch (final ProcessorStateException e) {
+                final String message = String.format(e.getMessage(), key);
+                throw new ProcessorStateException(message, e);
+            }
+        }
+
+        @Override
+        public <R> QueryResult<R> query(final Query<R> query,
+                                        final PositionBound positionBound,
+                                        final QueryConfig config) {
+            final long start = time.nanoseconds();
+            final QueryResult<R> result = wrapped().query(query, positionBound, config);
+            if (config.isCollectExecutionInfo()) {
+                result.addExecutionInfo(
+                    "Handled in " + getClass() + " in " + (time.nanoseconds() - start) + "ns");
+            }
+            // do not convert query or return types to/from inner bytes store to user-friendly types
```

Comment:
This design choice has implications -- I'm curious to hear reviewer thoughts.

The existing MeteredKeyValueStore and MeteredTimestampedKeyValueStore both contain logic to assist in translating `KeyQuery<K, V>` and `RangeQuery<K, V>` types to queries understandable by the underlying bytes store, and to deserialize the returned bytes result back for the original `K` and `V` result types. The return type of these queries for MeteredKeyValueStore and MeteredTimestampedKeyValueStore is `V` for the key query, and `KeyValueIterator<K, V>` for the range query. We'll want to introduce new return types for versioned stores, perhaps `VersionedRecord<V>` for the key query and `Iterator<VersionedRecord<V>>` for the range query. I'd like to postpone this decision-making for a follow-up KIP, but that means not assisting in translating to/from the inner bytes store for now.

So we have three options in the meantime:
1. Reject all IQv2 queries, since RocksDBVersionedKeyValueStore doesn't support IQv2 yet anyway. I don't think we should do this, because then users who create their own VersionedKeyValueStore implementations won't be able to issue IQv2 queries either.
2. Allow IQv2 queries as a pass-through at the metered layer. This is what the PR currently does. It means that users who implement their own VersionedKeyValueStores will be able to issue IQv2 queries, but also has backwards compatibility implications for when we introduce type serialization/deserialization support for key and range queries in the future. The behavior at this metered store layer would suddenly change for those query types.
3. Reject KeyQuery and RangeQuery types in order to reserve them for future behaviors, and pass-through all other query types. This avoids compatibility concerns for the future, while still allowing users to issue IQv2 queries to their own VersionedKeyValueStore implementations in the meantime.

I'm curious to know whether there's existing precedent / discussion on this compatibility concern for introducing custom logic at the metered layer for handling specific types of IQv2 queries.

Reply:
Hm, I think we might be talking about slightly different things. Here's my current understanding of how this works for the existing key-value stores today:
* The user passes a `KeyValueStore<Bytes, byte[]>` implementation (via supplier/materializer) which is used as the innermost store. This store only knows how to serve queries from bytes, and doesn't know anything about what the original key or value types are.
* The user's store gets wrapped inside some extra layers, including the metered layer which knows how to serialize the actual key and value types to bytes, and deserialize back.
* When the user issues a IQv2 query to the store, the query hits the outer store (i.e., metered layer), and is passed to the inner layers from there.
* What this means is that when a user issues an IQv2 `KeyQuery`, the key that they pass is the actual value type, and not bytes, even though the innermost store implementation they provided only knows about bytes. `MeteredKeyValueStore` has [logic](https://github.com/apache/kafka/blob/069ce59e1e33f47c000d8cdc247851f2e0a82154/streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java#L297) for serializing the key in the `KeyQuery` to bytes, and passing that to the inner store, and also deserializing the bytes from the result into the actual value type before returning the result to the user. The same thing happens with `RangeQuery`, but these are the only two query types which `MeteredKeyValueStore` provides serialization/deserialization assistance for. All other query types are direct pass-throughs to the inner store.

In order to provide the same convenience for users issuing IQv2 requests to versioned stores, `MeteredVersionedKeyValueStore` should also assist in serializing/deserializing `KeyQuery` and `RangeQuery` for users, so that they can issue `KeyQuery<K, V>` instead of `KeyQuery<Bytes, byte[]>` to their inner store, but I don't want to add this support now as part of KIP-889, since it should have its own KIP. So if a user wants to issue key queries to their custom versioned store implementation today, they will have to use `KeyQuery<Bytes, byte[]>`. But then later when we do add support for `KeyQuery<K, V>` at the metered store layer, users will have to stop using `KeyQuery<Bytes, byte[]>` and instead start using `KeyQuery<K, V>` -- this is the compatibility concern I have called out above. Does this extra context help?

Reply:
I spoke to @vvcephei who confirmed that allowing IQv2 queries to pass through the metered layer in this PR and then later following up with an in-built `KeyQuery` and/or `RangeQuery` implementation (to assist with serialization/deserialization of results) would be a breaking change that should be avoided. I will update this PR to reserve the `KeyQuery` and `RangeQuery` query types instead, i.e., throw an exception if those queries are issued while allowing other types of queries (e.g., user custom queries) to pass through.